### ISSUE TITLE
When loading save games (namely campaign missions), the mission specific track doesn't play

### ISSUE THREAD

**Issue by SuperSajuuk** [ncblakely/CataclysmEngineTweaks#7]
This has been an issue with Cataclysm/Emergence since it launched, and was hoping this mod could fix it.

Basically, to explain this bug report, all of the campaign missions have music tracks that play in those missions (ie an ambience track). However, due to a bug somewhere in the game engine, if you load a save game for any of these missions -- including ones you have created yourself, only the very generic music from Mission 01 is played over and over, which gets irritating after a while.

Of course, if the mission scripts a change in music (usually from going into battle, and then out of it), then this will be honoured by the mission parameters.

Not sure if this can be fixed by a mod like this, but if it can be, that would be a good fix to include. :)

NB: This bug doesn't apply in multiplayer, as players can use the , or . buttons to cycle the track to whatever they want to hear.

PS: This bug doesn't really exist in Homeworld 1, as the game properly loads the correct track to play in all the missions (although in that game, there is a minor bug whereby if you load a save game taken while a battle is underway and the associated battle track is playing, it may not always revert to the mission specific track: eg Mission 12 Galactic Core has this issue). For whatever reason, Cataclysm/Emergence broke the logic, so very often it falls back to generic ambience and not the mission-specific ambience.

**Comment by ncblakely** [comment]
Should be doable. I'll take a look. Thanks for the report.

**Comment by Talthos** [comment]
Worth noting is that, in Homeworld 1, a mission's baseline music track 'seemed' to be controlled by the `SongNumber` entry in the corresponding `.mission` file. And the number ordering seems to start at `0`, which is something to keep in mind if you've extracted the HW1 music tracks via Penesound with the default numbering settings in place.

Example: The very first HW1 mission uses `SongNumber    10`, which corresponds to `11.aifr` from a default Penesound music dump.

But in Cataclysm, for some reason, every `.mission` file uses `SongNumber    0`, which corresponds to `01.aifr` from a default Penesound music dump.

~Changing this value in the `.mission` file and substituting it (whether in a modified `update.big` or otherwise) doesn't seem to have any effect on in-game music behaviour upon loading a saved file in Cataclysm.~

Edit: Okay, I was mistaken. I must have misconfigured something during the initial tests I did before. I tested again just now, and changing the `SongNumber` entry in the `.mission` file _does_ work for Cataclysm.

**Comment by ncblakely** [comment]
This was a weird one, but I think it's fixed now. Cataclysm added some new code in the loading functions that erased the current track number after loading it from the save game, but before calling the music player.

I was about to commit something way more complex that involved writing extra data to the save file to save the track number before I realized it was already being saved :). Maybe I'll be able to use that code eventually...

This fix won't retroactively repair games that were saved with track number 0 active but it should be fixed after the next save with a different track playing.

I'll create a new release with this fix later this week after I have some more time to test it.

**Comment by SuperSajuuk** [comment]
> This was a weird one, but I think it's fixed now. Cataclysm added some new code in the loading functions that erased the current track number after loading it from the save game, but before calling the music player.
> 
> I was about to commit something way more complex that involved writing extra data to the save file to save the track number before I realized it was already being saved :). Maybe I'll be able to use that code eventually...
> 
> This fix won't retroactively repair games that were saved with track number 0 active but it should be fixed after the next save with a different track playing.
> 
> I'll create a new release with this fix later this week after I have some more time to test it.

@ncblakely great news! I'd love to download your engine fixes, but I only really want the music fix and none of the rest: will it be possible to have a config.ini or similar for the fixes so we can toggle on or off the fixes that we'd like to use? The rest of the fixes are great, but I personally just want to keep the vanilla experience, sans this one specific issue 👍 

I can make a separate issue ticket so that you can reference it in a commit if you end up adding this. :)

EDIT: Didn't see that there is an ini already! If I wanted to just add this specific fix for my game, what would i need to configure in the INI to make sure nothing else is changed?

**Comment by ncblakely** [comment]
Yeah, you can disable everything else. The .ini is pre-configured for higher spec machines, but if you set UpdateRateShift to 0, disable high detail mode and the new renderer selection, this bug fix should be the only thing left active.

I plan to keep everything configurable apart from fixes for obvious bugs going forwards.

You'll need v0.14 or compile from the source for the music fix (which I haven't published yet, but will this weekend).

**Comment by SuperSajuuk** [comment]
> Yeah, you can disable everything else. The .ini is pre-configured for higher spec machines, but if you set UpdateRateShift to 0, disable high detail mode and the new renderer selection, this bug fix should be the only thing left active.
> 
> I plan to keep everything configurable apart from fixes for obvious bugs going forwards.
> 
> You'll need v0.14 or compile from the source for the music fix (which I haven't published yet, but will this weekend).

Sounds good, thanks! Yeah, no need to toggle off bug fixes (we all love having bug fixes to improve our experience!!), but any enhancement that changes some aspect of the game definitely should be a toggle. 👍

Will keep an eye out for the next version and will download the engine fixes mod :)

**Comment by SuperSajuuk** [comment]
@ncblakely just downloaded the version with the fix and it looks like it _does_ retroactively apply the music patch to previously saved games. I loaded up Mission 12's "end of mission" save (which I saved without the CET patches mod in use) and the correct track automatically loaded without needing to save again. Had I not used this mod, I would have had the generic music track being played and not the mission specific one. I also took the opportunity to enable the NewRendererOptions so I don't have to deal with registry edits any more [didn't need SafeGL thankfully]. Good work!! :)
